# III. SQL-анализ и продуктовые метрики

**Цель ноутбука:**

- рассчитать ключевые продуктовые, поведенческие и финансовые метрики,
- изучить пользовательские сегменты и когорты,
- оценить активность и выручку сервиса.

### Импорт библиотек и подключение к базе данных

In [1]:
import pandas as pd
import os

from dotenv import load_dotenv
from urllib.parse import quote_plus

# импорт модуля для создания соединения с базой
from sqlalchemy import create_engine

In [2]:
# настройка отображения SQL-результатов в виде таблицы

import prettytable
prettytable.DEFAULT = prettytable.TableStyle

In [3]:
# загрузим расширение ipython-sql для выполнения SQL-запросов прямо в ноутбуке
%load_ext sql

Подключимся к PostgreSQL из ноутбука:

In [4]:
load_dotenv()

user = os.getenv("DB_USER")
password = quote_plus(os.getenv("DB_PASS"))
host = os.getenv("DB_HOST", "localhost")
port = int(os.getenv("DB_PORT", 5432))
database = os.getenv("DB_NAME")

# создадим соединение
engine = create_engine(f"postgresql+psycopg2://{user}:{password}@{host}:{port}/{database}")
conn_str = f"postgresql+psycopg2://{user}:{password}@{host}:{port}/{database}"

# проверим соединение
with engine.connect() as conn:
    print("Успешное подключение к PostgreSQL")

Успешное подключение к PostgreSQL


In [5]:
data_path = "../data/cleaned/kicksharing_clean.parquet"

df = pd.read_parquet(data_path)

df.to_sql("rides", engine, index=False, if_exists="replace")

print("Таблица rides успешно загружена в PostgreSQL")

Таблица rides успешно загружена в PostgreSQL


In [6]:
# подключимся к созданной базе данных через ipython-sql
conn_str = f"postgresql+psycopg2://{user}:{password}@{host}:{port}/{database}"
%sql $conn_str

Проверим работу SQL-запроса:

In [7]:
%sql SELECT * FROM rides LIMIT 5;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
5 rows affected.


[(252614779, 517007040, 'M', 26, 'GRD', 'UNM', 'Г МОСКВА', 7.99, 50.0, 300.0, 'SL', 0.8056608, datetime.datetime(2024, 4, 17, 16, 45, 16), datetime.datetime(2024, 4, 17, 16, 45, 20, 641980), datetime.datetime(2024, 4, 17, 16, 51, 26, 403392), 3, datetime.datetime(2024, 4, 17, 16, 45, 20, 641980), 97.94, 0.0, 0.0, 6.0960236, 7.9297013, 0.0, 0.0, 16, 2, 4),
 (252613066, 297828357, 'M', 31, 'Unknown', 'UNM', 'ХАНТЫ-МАНСИЙСКИЙ АВТОНОМНЫЙ ОКРУГ - ЮГРА АО', 7.99, 50.0, 300.0, 'SL', 0.23786059, datetime.datetime(2024, 4, 22, 15, 46, 4), datetime.datetime(2024, 4, 22, 15, 46, 10, 62458), datetime.datetime(2024, 4, 22, 15, 55, 4, 11996), 3, datetime.datetime(2024, 4, 22, 15, 46, 10, 62458), 121.91, 1.0, 0.0, 8.899159, 1.603706, 0.008202773, 0.0, 15, 0, 4),
 (252621809, 145875613, 'M', 35, 'GRD', 'DIV', 'Г МОСКВА', 7.99, 50.0, 300.0, 'SL', 0.28209627, datetime.datetime(2024, 4, 22, 15, 46, 53), datetime.datetime(2024, 4, 22, 15, 47, 0, 283797), datetime.datetime(2024, 4, 22, 15, 56, 28, 643724), 3, datetime.datetime(2024, 4, 22, 15, 47, 0, 283797), 129.9, 1.0, 0.0, 9.472666, 1.7868018, 0.00769823, 0.0, 15, 0, 4),
 (252632392, 328153536, 'M', 25, 'Unknown', 'Unknown', 'Г САНКТ-ПЕТЕРБУРГ', 6.99, 30.0, 300.0, 'SL', 0.3031659, datetime.datetime(2024, 4, 23, 6, 53, 41), datetime.datetime(2024, 4, 23, 6, 53, 45, 475088), datetime.datetime(2024, 4, 23, 6, 56, 59, 51051), 3, datetime.datetime(2024, 4, 23, 6, 53, 45, 475088), 57.96, 0.0, 0.0, 3.2262661, 5.6380825, 0.0, 0.0, 6, 1, 4),
 (252633271, 517007040, 'M', 26, 'GRD', 'UNM', 'Г МОСКВА', 7.49, 50.0, 300.0, 'SL', 0.8227758, datetime.datetime(2024, 4, 23, 9, 46, 17), datetime.datetime(2024, 4, 23, 9, 46, 22, 8459), datetime.datetime(2024, 4, 23, 9, 50, 38, 41837), 3, datetime.datetime(2024, 4, 23, 9, 46, 22, 8459), 87.45, 0.0, 0.0, 4.267223, 11.568776, 0.0, 0.0, 9, 1, 4)]

### Базовые метрики

##### Количество поездок и пользователей

In [8]:
%%sql
SELECT
    COUNT(DISTINCT order_rk) as total_rides,
    COUNT(DISTINCT party_rk) as total_users
FROM rides;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
1 rows affected.


[(352282, 57170)]

##### Период поездок

In [9]:
%%sql
SELECT
    MIN(book_start_dttm) as start,
    MAX(book_start_dttm) as end
FROM rides;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
1 rows affected.


[(datetime.datetime(2024, 4, 17, 16, 45, 20, 641980), datetime.datetime(2024, 10, 21, 17, 18, 7, 715000))]

- Поездки с 2024-04-17 по 2024-10-21

##### Статистики по поездкам

In [10]:
%%sql
SELECT
    ROUND(AVG(distance_km)::NUMERIC, 3) as avg_distance_km,
    PERCENTILE_CONT(0.5) WITHIN GROUP (ORDER BY distance_km) AS median_distance_km,
    ROUND(AVG(trip_duration_min)::NUMERIC, 2) as avg_trip_duration,
    PERCENTILE_CONT(0.5) WITHIN GROUP (ORDER BY trip_duration_min) AS median_trip_duration,
    ROUND(AVG(avg_speed_kmh)::NUMERIC, 2) as avg_speed_kmh,
    PERCENTILE_CONT(0.5) WITHIN GROUP (ORDER BY avg_speed_kmh) AS median_speed_kmh
FROM rides;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
1 rows affected.


[(Decimal('2.500'), 1.8028190732002258, Decimal('11.69'), 8.140203475952148, Decimal('13.74'), 13.130886554718018)]

- Средняя дистанция: 2,5 км
- Медианная дистанция: 1,8 км
- Средняя длительность: 11,7 мин
- Медианная длительность: 8,1 мин
- Средняя скорость: 13,7 км/ч
- Медианная скорость: 13,1 км/ч

_Среднее больше медианы, значит распределение скошено вправо._

##### TPU (trips per user)

In [11]:
%%sql
SELECT ROUND(COUNT(DISTINCT order_rk) / COUNT(DISTINCT party_rk)::NUMERIC, 2) as tpu FROM rides;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
1 rows affected.


[(Decimal('6.16'),)]

- Поездок на пользователя: 6.16

##### Количество поездок с кэшбэком

In [12]:
%%sql
WITH cashback_rides as (
SELECT
    COUNT(order_rk) FILTER(WHERE loyalty_accrual_rub_amt > 0) as rides_with_rub_cashback,
    COUNT(order_rk) FILTER(WHERE loyalty_accrual_bns_amt > 0) as rides_with_bns_cashback,
    COUNT(order_rk) FILTER(WHERE loyalty_accrual_rub_amt = 0 AND loyalty_accrual_bns_amt = 0) as rides_without_cashback,
    COUNT(order_rk) as total_rides
FROM rides
)

SELECT
    total_rides,
    rides_with_rub_cashback, ROUND((rides_with_rub_cashback * 1.0 / total_rides)::NUMERIC, 3) as rides_with_rub_cashback_share,
    rides_with_bns_cashback, ROUND((rides_with_bns_cashback * 1.0 / total_rides)::NUMERIC, 3) as rides_with_bns_cashback_share,
    rides_without_cashback, ROUND((rides_without_cashback * 1.0 / total_rides)::NUMERIC, 3) as rides_without_cashback_share
FROM cashback_rides;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
1 rows affected.


[(352282, 334094, Decimal('0.948'), 17450, Decimal('0.050'), 739, Decimal('0.002'))]

- Поездки с кэшбэком в рублях: 94,8%
- Поездки с кэшбэком в бонусах (использовалась кредитная карта): 5%
- Поездки без кэшбэка: 0,2%

### Продуктовые метрики

##### DAU

In [13]:
%%sql
SELECT
    DATE(local_book_start_dttm) as date,
    COUNT(DISTINCT party_rk) as dau
FROM rides
WHERE DATE(local_book_start_dttm) BETWEEN '2024-08-01' AND '2024-08-10'
GROUP BY date
ORDER BY date
LIMIT 10;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
10 rows affected.


[(datetime.date(2024, 8, 1), 1659),
 (datetime.date(2024, 8, 2), 2000),
 (datetime.date(2024, 8, 3), 1787),
 (datetime.date(2024, 8, 4), 1687),
 (datetime.date(2024, 8, 5), 1996),
 (datetime.date(2024, 8, 6), 2029),
 (datetime.date(2024, 8, 7), 2072),
 (datetime.date(2024, 8, 8), 1994),
 (datetime.date(2024, 8, 9), 2300),
 (datetime.date(2024, 8, 10), 1782)]

##### Average DAU

In [14]:
%%sql
WITH dau_metrics as (
    SELECT
        DATE(local_book_start_dttm) as date,
        COUNT(DISTINCT party_rk) as dau
    FROM rides
    GROUP BY date
    ORDER BY date
)

SELECT ROUND(AVG(dau), 0) as average_dau FROM dau_metrics;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
1 rows affected.


[(Decimal('1398'),)]

- Среднее количество ежедневных активных пользователей за весь период: 1398

##### WAU

In [15]:
%%sql
SELECT
    DATE_TRUNC('week', local_book_start_dttm)::DATE as week,
    COUNT(DISTINCT party_rk) as wau
FROM rides
GROUP BY week
ORDER BY week
LIMIT 10;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
10 rows affected.


[(datetime.date(2024, 4, 15), 1),
 (datetime.date(2024, 4, 22), 19),
 (datetime.date(2024, 4, 29), 27),
 (datetime.date(2024, 5, 6), 28),
 (datetime.date(2024, 5, 13), 350),
 (datetime.date(2024, 5, 20), 1031),
 (datetime.date(2024, 5, 27), 1616),
 (datetime.date(2024, 6, 3), 2505),
 (datetime.date(2024, 6, 10), 4355),
 (datetime.date(2024, 6, 17), 4655)]

##### Average WAU

In [16]:
%%sql
WITH wau_metrics as (
    SELECT
        DATE_TRUNC('week', local_book_start_dttm)::DATE as week,
        COUNT(DISTINCT party_rk) as wau
    FROM rides
    GROUP BY week
    ORDER BY week
)

SELECT ROUND(AVG(wau), 0) as average_wau FROM wau_metrics;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
1 rows affected.


[(Decimal('5537'),)]

- Среднее количество еженедельных активных пользователей за весь период: 5537

##### MAU

In [17]:
%%sql
SELECT
    DATE_TRUNC('month', local_book_start_dttm)::DATE AS month_d,
    COUNT(DISTINCT party_rk) as mau
FROM rides
GROUP BY month_d
ORDER BY month_d
LIMIT 10;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
7 rows affected.


[(datetime.date(2024, 4, 1), 23),
 (datetime.date(2024, 5, 1), 2130),
 (datetime.date(2024, 6, 1), 11632),
 (datetime.date(2024, 7, 1), 17529),
 (datetime.date(2024, 8, 1), 25196),
 (datetime.date(2024, 9, 1), 24939),
 (datetime.date(2024, 10, 1), 13346)]

##### Average MAU

In [18]:
%%sql
WITH mau_metrics as (
    SELECT
        DATE_TRUNC('month', local_book_start_dttm)::DATE AS month_d,
        COUNT(DISTINCT party_rk) as mau
    FROM rides
    GROUP BY month_d
    ORDER BY month_d
)

SELECT ROUND(AVG(mau), 0) as average_mau FROM mau_metrics;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
1 rows affected.


[(Decimal('13542'),)]

- Среднее количество ежемесячных активных пользователей за весь период: 13542

### Финансовые метрики

##### Revenue

In [19]:
%%sql
SELECT ROUND(SUM(nominal_price_rub_amt)::NUMERIC, 2) as revenue_rub FROM rides;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
1 rows affected.


[(Decimal('45791400.00'),)]

- Выручка сервиса: 45,7 млн. рублей

##### ARPU

In [20]:
%%sql
SELECT ROUND((SUM(nominal_price_rub_amt) / COUNT(DISTINCT party_rk))::NUMERIC, 2) as arpu FROM rides;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
1 rows affected.


[(Decimal('800.54'),)]

- ARPU: 800,5 рублей

##### Average check

In [21]:
%%sql
SELECT
    ROUND(AVG(nominal_price_rub_amt)::NUMERIC, 2) as avg_price_per_ride_rub,
    ROUND((PERCENTILE_CONT(0.5) WITHIN GROUP (ORDER BY nominal_price_rub_amt))::NUMERIC, 2) as median_price_per_ride_rub
FROM rides;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
1 rows affected.


[(Decimal('130.05'), Decimal('105.93'))]

- Средняя стоимость поездки: 130,05 рублей
- Медианная стоимость поездки: 105,93 рублей

_Это говорит о наличии дорогих поездок (правый хвост распределения)._

##### Average cashback

In [22]:
%%sql
SELECT
    ROUND(AVG(loyalty_accrual_rub_amt) FILTER (WHERE loyalty_accrual_rub_amt > 0) :: NUMERIC, 2) as avg_rub_cashback,
    ROUND(AVG(loyalty_accrual_bns_amt) FILTER (WHERE loyalty_accrual_bns_amt > 0) :: NUMERIC, 2) as avg_bns_cashback
FROM rides;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
1 rows affected.


[(Decimal('34.65'), Decimal('57.86'))]

- Средний кэшбэк для поездок с оплатой по дебетовой карте: 34,65 рубля
- Средний кэшбэк для поездок с оплатой по кредитной карте: 57,86 бонусов

##### Средний чек по типу оплаты

In [23]:
%%sql
WITH type_cashback as (
    SELECT
        order_rk,
        CASE
            WHEN loyalty_accrual_rub_amt > 0 THEN 'debet'
            WHEN loyalty_accrual_bns_amt > 0 THEN 'credit'
            ELSE 'unknown'
        END as pay_type
    FROM rides
)

SELECT t.pay_type, AVG(nominal_price_rub_amt) as avg_price_per_ride
FROM rides r
    INNER JOIN type_cashback t USING(order_rk)
GROUP BY t.pay_type;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
3 rows affected.


[('unknown', 119.43132654223616),
 ('credit', 131.2397438939427),
 ('debet', 130.01285005547066)]

_Тип оплаты не влияет на среднюю стоимость поездки._

##### Cashback share

In [24]:
%%sql
SELECT
    SUM(loyalty_accrual_rub_amt) as sum_cashback_rub,
    ROUND((SUM(loyalty_accrual_rub_amt) / SUM(nominal_price_rub_amt))::NUMERIC, 2) as cashback_share_rub,
    SUM(loyalty_accrual_bns_amt) as sum_cashback_bns,
    ROUND((SUM(loyalty_accrual_bns_amt) / SUM(nominal_price_rub_amt))::NUMERIC, 2) as cashback_share_bns
FROM rides;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
1 rows affected.


[(11575729.0, Decimal('0.25'), 1009699.8, Decimal('0.02'))]

- Доля выплаченного кэшбэка в рублях к выручке: 25%
- Доля выплаченного кэшбэка в бонусах к выручке: 2%

### Сегментация пользователей

##### По активности

In [25]:
%%sql
WITH users_rides as (
    SELECT
        party_rk,
        COUNT(order_rk) as rides,
        SUM(nominal_price_rub_amt) as revenue
    FROM rides
    GROUP BY party_rk
),
segm as (
    SELECT *,
        CASE
            WHEN rides = 1 THEN 'one-time (1)'
            WHEN rides BETWEEN 2 AND 3 THEN 'casual (2-3)'
            WHEN rides BETWEEN 4 AND 10 THEN 'regular (4-10)'
            ELSE 'heavy (10+)'
        END as segment
    FROM users_rides
),
total_stats as (
    SELECT COUNT(DISTINCT party_rk) as total_users, SUM(nominal_price_rub_amt) as total_revenue FROM rides
)

SELECT
    segment,
    COUNT(party_rk) as users,
    ROUND(COUNT(party_rk) / (SELECT total_users FROM total_stats)::NUMERIC, 3) as users_share,
    SUM(rides) as total_rides_per_segment,
    SUM(revenue) as revenue,
    ROUND((SUM(revenue) / (SELECT total_revenue FROM total_stats))::NUMERIC, 3) as revenue_share,
    ROUND(AVG(rides), 2) as avg_rides_per_segment
FROM segm
GROUP BY segment
LIMIT 10;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
4 rows affected.


[('heavy (10+)', 7728, Decimal('0.135'), Decimal('223445'), 26670478.0, Decimal('0.583'), Decimal('28.91')),
 ('regular (4-10)', 11747, Decimal('0.205'), Decimal('70993'), 9572706.0, Decimal('0.209'), Decimal('6.04')),
 ('one-time (1)', 22892, Decimal('0.400'), Decimal('22892'), 4230162.0, Decimal('0.092'), Decimal('1.00')),
 ('casual (2-3)', 14803, Decimal('0.259'), Decimal('34952'), 5341609.5, Decimal('0.117'), Decimal('2.36'))]

- Heavy пользователи (с 10+ поездками) дают **58,3% всей выручки**, хотя их только 13,5%.
- One-time пользователи (их 40%) — дают лишь 9,2% выручки.

##### Топ-10 прибыльных пользователей

In [26]:
%%sql
SELECT
    party_rk,
    COUNT(order_rk) as rides,
    SUM(nominal_price_rub_amt) as revenue,
    RANK() OVER(ORDER BY SUM(nominal_price_rub_amt) DESC) as rank
FROM rides
GROUP BY party_rk
LIMIT 10;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
10 rows affected.


[(386134592, 335, 53965.418, 1),
 (759042976, 273, 50547.777, 2),
 (910211081, 333, 47343.082, 3),
 (284170812, 191, 36849.812, 4),
 (636557947, 300, 34166.72, 5),
 (231464761, 285, 33580.574, 6),
 (666474893, 186, 31686.42, 7),
 (528704373, 156, 31311.553, 8),
 (391309281, 299, 29861.793, 9),
 (976560193, 166, 29747.06, 10)]

Активные пользователи и правда активные — стоит рассмотреть программы лояльности или таргет-акции для них.

##### По полу

In [27]:
%%sql
SELECT
    gender_cd,
    COUNT(DISTINCT party_rk) as users,
    COUNT(order_rk) as total_rides,
    ROUND(AVG(distance_km)::NUMERIC, 3) as avg_distance_km,
    ROUND(AVG(nominal_price_rub_amt)::NUMERIC, 2) as avg_price,
    ROUND(AVG(trip_duration_min)::NUMERIC, 2) as avg_trip_duration_min,
    ROUND(AVG(cashback_rate_rub)::NUMERIC, 2) as avg_cashback_rub,
    ROUND(AVG(cashback_rate_bns)::NUMERIC, 2) as avg_cashback_bns
FROM rides
GROUP BY gender_cd
ORDER BY 2 DESC;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
3 rows affected.


[('M', 46117, 297850, Decimal('2.499'), Decimal('128.13'), Decimal('11.46'), Decimal('0.23'), Decimal('0.03')),
 ('F', 10622, 51827, Decimal('2.537'), Decimal('141.62'), Decimal('13.08'), Decimal('0.25'), Decimal('0.02')),
 ('U', 431, 2605, Decimal('1.891'), Decimal('119.43'), Decimal('9.94'), Decimal('0.23'), Decimal('0.00'))]

Средняя длительность поездки и средняя стоимость поездки у женщин выше, чем у мужчин при относительно равной средней дистанции.

##### По количеству поездок на пользователя

In [28]:
%%sql
SELECT gender_cd, ROUND(COUNT(order_rk)::NUMERIC / COUNT(DISTINCT party_rk), 2) as tpu_by_gender
FROM rides
GROUP BY gender_cd;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
3 rows affected.


[('F', Decimal('4.88')), ('M', Decimal('6.46')), ('U', Decimal('6.04'))]

- Количество поездок на пользователя у мужчин: 6,46
- Количество поездок на пользователя у женщин: 4,88

##### По регионам

In [29]:
%%sql
SELECT
    lvn_state_nm as region,
    COUNT(*) as total_rides,
    COUNT(DISTINCT party_rk) as users,
    ROUND(AVG(distance_km)::NUMERIC, 3) as avg_distance_km,
    ROUND(AVG(nominal_price_rub_amt)::NUMERIC, 2) as avg_price,
    ROUND(AVG(trip_duration_min)::NUMERIC, 2) as avg_trip_duration_min,
    SUM(nominal_price_rub_amt) as revenue,
    RANK() OVER(ORDER BY SUM(nominal_price_rub_amt)::NUMERIC DESC) as rank
FROM rides
WHERE lvn_state_nm <> 'Unknown'
GROUP BY lvn_state_nm
LIMIT 10;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
10 rows affected.


[('Г МОСКВА', 81409, 10699, Decimal('2.115'), Decimal('132.77'), Decimal('10.01'), 10810632.0, 1),
 ('МОСКОВСКАЯ ОБЛ', 35666, 5398, Decimal('2.229'), Decimal('132.76'), Decimal('10.69'), 4735776.5, 2),
 ('Г САНКТ-ПЕТЕРБУРГ', 34127, 5029, Decimal('2.688'), Decimal('128.52'), Decimal('10.92'), 4386705.5, 3),
 ('СВЕРДЛОВСКАЯ ОБЛ', 25879, 2962, Decimal('2.354'), Decimal('106.38'), Decimal('11.46'), 2753187.0, 4),
 ('КРАСНОДАРСКИЙ КРАЙ', 14442, 1972, Decimal('3.052'), Decimal('129.88'), Decimal('12.78'), 1875790.5, 5),
 ('РЕСП ТАТАРСТАН', 8670, 1608, Decimal('2.733'), Decimal('117.12'), Decimal('13.10'), 1015384.9, 6),
 ('РЕСП БАШКОРТОСТАН', 7441, 1411, Decimal('2.358'), Decimal('125.23'), Decimal('12.94'), 931808.25, 7),
 ('НИЖЕГОРОДСКАЯ ОБЛ', 6700, 1306, Decimal('2.403'), Decimal('131.03'), Decimal('12.39'), 877915.4, 8),
 ('РЕСП УДМУРТСКАЯ', 4918, 861, Decimal('3.559'), Decimal('131.50'), Decimal('14.20'), 646706.25, 9),
 ('НОВОСИБИРСКАЯ ОБЛ', 6106, 1037, Decimal('2.203'), Decimal('104.50'), Decimal('11.98'), 638088.25, 10)]

##### По моделям самокатов

In [30]:
%%sql
SELECT
    transport_model,
    COUNT(*) as total_rides,
    ROUND(AVG(distance_km)::NUMERIC, 3) as avg_distance_km,
    ROUND(AVG(nominal_price_rub_amt)::NUMERIC, 2) as avg_price,
    ROUND(AVG(trip_duration_min)::NUMERIC, 2) as avg_trip_duration_min,
    ROUND(AVG(cashback_rate_rub)::NUMERIC, 2) as avg_cashback_rub
FROM rides
GROUP BY transport_model
ORDER BY 2 DESC;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
13 rows affected.


[('SL', 217135, Decimal('2.511'), Decimal('132.10'), Decimal('11.06'), Decimal('0.23')),
 ('Max', 61050, Decimal('2.720'), Decimal('133.99'), Decimal('13.13'), Decimal('0.23')),
 ('E', 59440, Decimal('2.255'), Decimal('114.34'), Decimal('11.88'), Decimal('0.23')),
 ('Urban E-Bike', 3874, Decimal('2.694'), Decimal('132.31'), Decimal('11.19'), Decimal('0.27')),
 ('ES400', 3602, Decimal('2.459'), Decimal('140.68'), Decimal('14.29'), Decimal('0.26')),
 ('ES200', 2802, Decimal('2.102'), Decimal('145.18'), Decimal('13.41'), Decimal('0.24')),
 ('ES100', 2368, Decimal('2.191'), Decimal('159.62'), Decimal('16.57'), Decimal('0.28')),
 ('MK', 926, Decimal('2.496'), Decimal('157.62'), Decimal('16.49'), Decimal('0.22')),
 ('ES-08S', 705, Decimal('2.799'), Decimal('199.65'), Decimal('19.27'), Decimal('0.24')),
 ('Techno', 195, Decimal('1.971'), Decimal('149.14'), Decimal('13.59'), Decimal('0.26')),
 ('ES', 91, Decimal('2.386'), Decimal('183.10'), Decimal('20.23'), Decimal('0.31')),
 ('LB', 71, Decimal('3.125'), Decimal('176.26'), Decimal('18.82'), Decimal('0.22')),
 ('B1D', 23, Decimal('2.672'), Decimal('278.07'), Decimal('25.42'), Decimal('0.40'))]

### Когортный анализ

##### Когортный анализ по дате первой поездки (месяц первой поездки)

In [31]:
%%sql
WITH first_ride as (
    SELECT
        party_rk,
        MIN(DATE(local_book_start_dttm)) as first_date
    FROM rides
    GROUP BY party_rk
), cohort as (
    SELECT
        party_rk,
        DATE_TRUNC('month', first_date)::DATE as cohort_month
    FROM first_ride
), activity as (
    SELECT
        r.party_rk,
        DATE_TRUNC('month', r.local_book_start_dttm)::DATE as activity_month,
        c.cohort_month,
        EXTRACT(
            MONTH FROM age(DATE_TRUNC('month', r.local_book_start_dttm),
                           c.cohort_month)
        ) as cohort_age
    FROM rides r
    JOIN cohort c USING (party_rk)
), cohort_counts as (
    SELECT cohort_month, COUNT(DISTINCT party_rk) as cohort_size
    FROM activity
    WHERE cohort_age = 0
    GROUP BY cohort_month
)

SELECT
    a.cohort_month,
    a.cohort_age,
    a.users,
    ROUND(a.users::DECIMAL / c.cohort_size, 4) as retention
FROM (
    SELECT cohort_month, cohort_age, COUNT(DISTINCT party_rk) as users
    FROM activity
    GROUP BY cohort_month, cohort_age
) a
JOIN cohort_counts c USING (cohort_month)
ORDER BY cohort_month, cohort_age;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
28 rows affected.


[(datetime.date(2024, 4, 1), Decimal('0'), 23, Decimal('1.0000')),
 (datetime.date(2024, 4, 1), Decimal('1'), 12, Decimal('0.5217')),
 (datetime.date(2024, 4, 1), Decimal('2'), 11, Decimal('0.4783')),
 (datetime.date(2024, 4, 1), Decimal('3'), 16, Decimal('0.6957')),
 (datetime.date(2024, 4, 1), Decimal('4'), 11, Decimal('0.4783')),
 (datetime.date(2024, 4, 1), Decimal('5'), 13, Decimal('0.5652')),
 (datetime.date(2024, 4, 1), Decimal('6'), 7, Decimal('0.3043')),
 (datetime.date(2024, 5, 1), Decimal('0'), 2118, Decimal('1.0000')),
 (datetime.date(2024, 5, 1), Decimal('1'), 984, Decimal('0.4646')),
 (datetime.date(2024, 5, 1), Decimal('2'), 850, Decimal('0.4013')),
 (datetime.date(2024, 5, 1), Decimal('3'), 710, Decimal('0.3352')),
 (datetime.date(2024, 5, 1), Decimal('4'), 611, Decimal('0.2885')),
 (datetime.date(2024, 5, 1), Decimal('5'), 388, Decimal('0.1832')),
 (datetime.date(2024, 6, 1), Decimal('0'), 10637, Decimal('1.0000')),
 (datetime.date(2024, 6, 1), Decimal('1'), 5055, Decimal('0.4752')),
 (datetime.date(2024, 6, 1), Decimal('2'), 3846, Decimal('0.3616')),
 (datetime.date(2024, 6, 1), Decimal('3'), 3106, Decimal('0.2920')),
 (datetime.date(2024, 6, 1), Decimal('4'), 1915, Decimal('0.1800')),
 (datetime.date(2024, 7, 1), Decimal('0'), 11608, Decimal('1.0000')),
 (datetime.date(2024, 7, 1), Decimal('1'), 4459, Decimal('0.3841')),
 (datetime.date(2024, 7, 1), Decimal('2'), 3167, Decimal('0.2728')),
 (datetime.date(2024, 7, 1), Decimal('3'), 1880, Decimal('0.1620')),
 (datetime.date(2024, 8, 1), Decimal('0'), 16170, Decimal('1.0000')),
 (datetime.date(2024, 8, 1), Decimal('1'), 5228, Decimal('0.3233')),
 (datetime.date(2024, 8, 1), Decimal('2'), 2657, Decimal('0.1643')),
 (datetime.date(2024, 9, 1), Decimal('0'), 12814, Decimal('1.0000')),
 (datetime.date(2024, 9, 1), Decimal('1'), 2699, Decimal('0.2106')),
 (datetime.date(2024, 10, 1), Decimal('0'), 3800, Decimal('1.0000'))]

**Анализ:** в ранних когортах (май-июнь) почти половина пользователей возвращалась через месяц после регистрации. Начиная с июля, retention снизился до 38%, а в августовской когорте через месяц вернулось всего 32% пользователей. Тренд указывает на снижение вовлечённости новых пользователей с течением времени — необходимо усиление стимулов для повторной поездки: промокод может увеличить retention.

### Выводы

- Выручка распределена неравномерно — на heavy пользователей (10+ поездок) приходится больше половины дохода.
- Мужчины и женщины совершают разное количество поездок за период пользования сервисом, что подтверждает анализ из EDA.
- На 99,8% поездок был начислен кэшбэк либо в рублях, либо в бонусах в зависимости от типа оплаты.
- Стоимость поездки не зависит от типа оплаты.
- Постепенное снижение retention со временем указывает на необходимость дополнительных механик удержания пользователей.